File: Exception_handling_Patterns:Anthony_Argueta/ExceptionHandlingPatters_AnthonyArgueta/SourceCode/ruby/file.reader.rb
def read_file(filename)
    begin
        puts "opening file..."
        file = File.open(filename, "r") #raises error file not found
        begin
            puts "reading file..."
            data = file.read
            return data
        ensure
            puts 'closing file...'
            file.close if file
        end 
    rescue Errno::ENOENT => e
        puts "file not found: #{e.message}"
        raise
    rescue IOError => e
        puts "General exception: #{e.message}"
        raise
    end
end

def main
    begin
        read_file("does_not_exist.txt")
    rescue => e
        puts "propagated to main: #{e.message}"
    ensure
        puts "execution completed."
    end
end

main

File: Exception_handling_Patterns:Anthony_Argueta/ExceptionHandlingPatters_AnthonyArgueta/SourceCode/java/filereader.java
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

public class FileReader {

    public static String readFile(String filename) throws IOException {
        System.out.println("opening file...");
        FileReader reader = null;

        try {
            reader = new FileReader(filename);
            System.out.println("reading file...");

            StringBuilder sb = new StringBuilder();
            int ch;

            while ((ch = reader.read()) != -1) {
                sb.append((char) ch);
            }

            return sb.toString();`
        }

        finally {
            System.out.println("closing file...");
            if (reader != null) {
                reader.close();
            }
        }

    }

    public static void main(String[] args) {
        try {
            String text = readFile("does_not_exist.txt");
            System.out.println(text);
        }
        catch (FileNotFoundException e) {
            System.out.println("Error: File not found");
        }
        catch (IOException e) {
            System.out.println("Error: I/O error occurred" + e);
        }
        finally {
            System.out.println("Finished java program");
        }
    }
}